src/registry.rs

```diff
@@ -434,228 +445,574 @@ impl DependencyResolution {
+                    indexmap::map::Entry::Vacant(e) => {
+                        let url = find_url(self.config, Some(registry_name), self.urls)?;
+                        e.insert(Registry {
+                            client: Arc::new(create_client(self.config, url)?),
+                            packages: HashMap::new(),
+                            dependencies: Vec::new(),
+                            upserts: HashSet::new(),
+                        })
                     }
-                }
+                };
+
+                registry
+                    .add_dependency(name, package, locked, from_target)
+                    .await?;
+            }
+            metadata::Dependency::Local(p) => {
+                // A local path dependency, insert a resolution immediately
+                let res = DependencyResolution::Local(LocalResolution {
+                    name: name.to_string(),
+                    path: p.clone(),
+                });
+
+                let prev = if from_target {
+                    self.target_resolutions.insert(name.to_string(), res)
+                } else {
+                    self.resolutions.insert(name.to_string(), res)
+                };
+
+                assert!(prev.is_none());
             }
         }
 
-        return Ok(Some(resolution));
+        Ok(())
+    }
 
-        struct PackageEntry<'a> {
-            name: &'a str,
-            package: &'a metadata::RegistryPackage,
-            target: bool,
+    /// Resolve all dependencies.
+    ///
+    /// This will download all dependencies that are not already present in client storage.
+    ///
+    /// Returns a tuple of target dependency resolutions and component dependency resolutions.
+    pub async fn resolve(mut self) -> Result<(DependencyResolutionMap, DependencyResolutionMap)> {
+        self.update_and_download().await?;
+
+        for registry in self.registries.into_values() {
+            for dep in registry.dependencies {
+                let resolution = DependencyResolution::Registry(
+                    dep.resolution.expect("dependency should be resolved"),
+                );
+
+                let prev = if dep.from_target {
+                    self.target_resolutions
+                        .insert(dep.name.to_string(), resolution)
+                } else {
+                    self.resolutions.insert(dep.name.to_string(), resolution)
+                };
+
+                assert!(prev.is_none());
+            }
         }
 
-        struct RegistryEntry<'a> {
-            registry: Box<dyn Registry>,
-            packages: HashMap<&'a PackageId, Vec<PackageEntry<'a>>>,
-        }
+        Ok((self.target_resolutions, self.resolutions))
     }
 
-    /// Iterates over all dependencies of the package.
-    pub fn deps(&self) -> impl Iterator<Item = (&String, &DependencyResolution)> {
-        self.target_dependencies
+    async fn update_and_download(&mut self) -> Result<()> {
```

Comment:
Separate out update and download into their own functions that pipe data through.

Reply:
I've split this up a little bit better where the update and "download and resolve" steps feed state through.

Let me know if the change doesn't make sense.